### PR TITLE
Demo for YOLO model with postprocess layers

### PR DESCRIPTION
**Pull request by likholat**
Added post-processing to the model during conversion and excluded it from the demo.
requires https://github.com/openvinotoolkit/openvino/pull/6088

To convert model:
```
python3 converter.py --name yolo-v4-tf --mo /openvino/model-optimizer/mo.py  --add_mo_arg='--transformations_config=/openvino/model-optimizer/extensions/front/tf/yolo_v4.json'
```

**Performance:**
Launched on `Intel(R) Core(TM) i7-8700 CPU @ 3.20GHz`
Input video info:
```
File size : 5.68 MiB
Duration : 2 min 23 s
```
```
python3 object_detection_demo.py -m yolovX_model.xml -at yolo -i /path/to/video --no_show
```
YoloV3 results:
|Python demo | Original model  | Model with postprocess |
| --- | --- | --- |
| Latency | 485.2 ms | 511.5 ms |
| FPS |10.3 | 9.7|

|C++ demo | Original model  | Model with postprocess |
| --- | --- | --- |
| Latency | 483.4 ms | 492.0 ms|
| FPS | 10.3 | 10.2 |

YoloV4 results:
|Python demo | Original model  | Model with postprocess |
| --- | --- | --- |
| Latency | 1076.1 ms | 1125.7 ms|
| FPS | 4.6 |4.4|

### CODE REVIEW

**Comment by likholat** [comment]
https://github.com/openvinotoolkit/openvino/pull/6088#pullrequestreview-678824076

File: demos/common/python/models/yolo.py [hunk 1]
```diff
@@ -197,11 +224,13 @@ def _resize_detections_letterbox(detections, original_shape, resized_shape):
 
     def postprocess(self, outputs, meta):
         detections = list()
-
         for layer_name, out_blob in outputs.items():
-            layer_params = self.yolo_layer_params[layer_name]
-            out_blob.shape = layer_params[0]
-            detections += self._parse_yolo_region(out_blob, meta['resized_shape'], layer_params[1], self.threshold)
+            if 'YoloRegion' in layer_name:
```
Comment (dkurt) [inline demos/common/python/models/yolo.py:228]: Can we check layer type instead?

File: demos/common/python/models/yolo.py [hunk 2]
```diff
@@ -143,6 +143,32 @@ def _parse_yolo_region(predictions, input_size, params, threshold, multiple_labe
                                          confidence.item(), class_id.item()))
         return objects
 
+    # For model which includes postprocess
+    @staticmethod
+    def _parse_yolo_with_postprocess_region(predictions, threshold, multiple_labels=True):
```
Comment (dkurt) [inline demos/common/python/models/yolo.py:148]: Optionn multiple_labels is always True. Do we really need it?

File: demos/object_detection_demo/cpp/main.cpp [hunk 3]
```diff
@@ -381,6 +381,7 @@ int main(int argc, char *argv[]) {
         pipeline.waitForTotalCompletion();
         while (result = pipeline.getResult()) {
             cv::Mat outFrame = renderDetectionData(result->asRef<DetectionResult>(), palette);
+            cv::imwrite("cpp_res.jpg", outFrame);
```
Comment (dkurt) [inline demos/object_detection_demo/cpp/main.cpp:384]: Extra change

File: demos/common/cpp/models/src/detection_model_yolo.cpp [hunk 4]
```diff
@@ -91,8 +94,12 @@ std::unique_ptr<ResultBase> ModelYolo3::postprocess(InferenceResult & infResult)
     const auto& internalData = infResult.internalModelData->asRef<InternalImageModelData>();
 
     for (auto& output : infResult.outputsData) {
-        this->parseYOLOV3Output(output.first, output.second, netInputHeight, netInputWidth,
-            internalData.inputImgHeight, internalData.inputImgWidth, objects);
+        if (output.first.find("YoloRegion") != std::string::npos){
+            this->parseYOLOV3Output(output.first, output.second, netInputHeight, netInputWidth,
+                internalData.inputImgHeight, internalData.inputImgWidth, objects);
```
Comment (dkurt) [inline demos/common/cpp/models/src/detection_model_yolo.cpp:99]: Please fix alignment

File: demos/common/cpp/models/include/models/detection_model_yolo.h [hunk 5]
```diff
@@ -63,6 +63,9 @@ class ModelYolo3 : public DetectionModel {
         const unsigned long resized_im_h, const unsigned long resized_im_w, const unsigned long original_im_h,
         const unsigned long original_im_w, std::vector<DetectedObject>& objects);
 
+    void parseYOLOV3Output(const InferenceEngine::Blob::Ptr& blob,
+        const unsigned long original_im_h, const unsigned long original_im_w, std::vector<DetectedObject>& objects);
```
Comment (dkurt) [inline demos/common/cpp/models/include/models/detection_model_yolo.h:67]: Alignment

**Comment by dkurt** [comment]
Measure efficiency for both Python and C++ demos

File: demos/common/cpp/models/src/detection_model_yolo.cpp [hunk 6]
```diff
@@ -194,6 +204,43 @@ void ModelYolo3::parseYOLOV3Output(const std::string& output_name,
...
+    const float* output_blob = blob->buffer().as<PrecisionTrait<Precision::FP32>::value_type*>();
+
+    // --------------------------- Parsing YOLO output --------------------------------------------
+    for (int res_num = 0; res_num < detections_num; ++res_num) {
+        float object_score = output_blob[res_num * len_of_res + 4];
+        if (object_score < confidenceThreshold) {
+            continue;
+        }
+
+        for (int i = 5; i < len_of_res; ++i) {
```
Comment (dkurt) [inline demos/common/cpp/models/src/detection_model_yolo.cpp:222]: What is 5?

Comment (likholat) [reply demos/common/cpp/models/src/detection_model_yolo.cpp:222]: The first five elements are coordinates and object score, elements sixth to n are probabilities. Five is the index of the element from which the probabilities start.

File: demos/common/cpp/models/src/detection_model_yolo.cpp [hunk 7]
```diff
@@ -192,6 +201,43 @@ void ModelYolo3::parseYOLOV3Output(const std::string& output_name,
...
+                float width = output_blob[res_num * len_of_res + 2];
+                float height = output_blob[res_num * len_of_res + 3];
+
+                DetectedObject obj;
+                obj.x = (float)(x - width / 2) * original_im_w;
+                obj.y = (float)(y - height / 2) * original_im_h;
+                obj.width = (float)(width) * original_im_w;
+                obj.height = (float)(height) * original_im_h;
+                obj.confidence = class_probability;
+                obj.labelID = i - 5;
```
Comment (dkurt) [inline demos/common/cpp/models/src/detection_model_yolo.cpp:233]: Why we have to subtract 5?

File: demos/common/cpp/models/include/models/detection_model_yolo.h [hunk 8]
```diff
@@ -65,6 +65,9 @@ class ModelYolo : public DetectionModel {
         const unsigned long resized_im_h, const unsigned long resized_im_w, const unsigned long original_im_h,
         const unsigned long original_im_w, std::vector<DetectedObject>& objects);
 
+    void parseYOLOOutput(const InferenceEngine::Blob::Ptr& blob,
+                           const unsigned long original_im_h, const unsigned long original_im_w, std::vector<DetectedObject>& objects);
```
Comment (dkurt) [inline demos/common/cpp/models/include/models/detection_model_yolo.h:69]: please fix alignment

File: demos/common/cpp/models/src/detection_model_yolo.cpp [hunk 9]
```diff
@@ -137,6 +146,14 @@ void ModelYolo::parseYOLOOutput(const std::string& output_name,
     const unsigned long original_im_w,
     std::vector<DetectedObject>& objects) {
 
+    const int out_blob_h = static_cast<int>(blob->getTensorDesc().getDims()[2]);
+    const int out_blob_w = static_cast<int>(blob->getTensorDesc().getDims()[3]);
+    if (out_blob_h != out_blob_w) {
+        throw std::runtime_error("Invalid size of output " + output_name +
+            " It should be in NCHW layout and H should be equal to W. Current H = " + std::to_string(out_blob_h) +
+            ", current W = " + std::to_string(out_blob_w));
+    }
```
Comment (dkurt) [inline demos/common/cpp/models/src/detection_model_yolo.cpp:155]: Why do we need this limitation?

File: demos/common/cpp/models/src/detection_model_yolo.cpp [hunk 10]
```diff
@@ -214,6 +231,49 @@ int ModelYolo::calculateEntryIndex(int totalCells, int lcoords, int lclasses, in
...
+                float x = output_blob[res_num * len_of_res];
+                float y = output_blob[res_num * len_of_res + 1];
+                float width = output_blob[res_num * len_of_res + 2];
+                float height = output_blob[res_num * len_of_res + 3];
+
+                DetectedObject obj;
+                obj.x = (float)(x - width / 2) * original_im_w;
+                obj.y = (float)(y - height / 2) * original_im_h;
+                obj.width = (float)(width) * original_im_w;
+                obj.height = (float)(height) * original_im_h;
```
Comment (dkurt) [inline demos/common/cpp/models/src/detection_model_yolo.cpp:261]:
```
                obj.width = width * original_im_w;
                obj.height = height * original_im_h;
                obj.x = x * original_im_w - obj.width / 2;
                obj.y = y * original_im_h - obj.height / 2;
```

**Comment by likholat** [comment]
@eizamaliev, please review

**Comment by likholat** [comment]
@eizamaliev, the output of model with postprocessing is `num_of_predictions x 85`, where 85 is x, y, width, height, object_probability and 80 classes

**Comment by likholat** [comment]
@eizamaliev, should we keep the Yolo format or would it be preferable to change the output to a `1,1,7,N` format?

**Comment by vladimir-dudnik** [comment]
@likholat precommit check failed

20:30:07  running miscellaneous checks...
20:30:07  demos/common/cpp/models/src/detection_model_yolo.cpp:239: trailing whitespace.
20:30:07  +        const float* max_probability = std::max_element(&output_blob[res_num * len_of_res + 5],

**Comment by dkurt** [comment]
@likholat, please provide new benchmark measurements considering https://github.com/openvinotoolkit/open_model_zoo/pull/2725

**Comment by vladimir-dudnik** [comment]
20:30:07  demos/common/cpp/models/src/detection_model_yolo.cpp:239: trailing whitespace.
20:37:10    c:\jenkins\workspace\demos_windows10ce29c7ab\src\demos\common\cpp\models\src\detection_model_yolo.cpp(246): warning C4244: 'initializing': conversion from '__int64' to 'int', possible loss of data

**Comment by vladimir-dudnik** [comment]
yolo-v4-tf
00:23:35  [ ERROR ] Can't find output layer with name StatefulPartitionedCall/model/conv2d_101/BiasAdd/Add

yolo-v4-tiny-tf
00:23:37  [ ERROR ] Can't find output layer with name conv2d_17/BiasAdd/Add

**Comment by vladimir-dudnik** [comment]
Jenkins please retry a build

**Pull request closed without merging**